pkg/filesystem/subvolume.go

```diff
@@ -0,0 +1,183 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subvolume
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/rook/kubectl-rook-ceph/pkg/exec"
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+	"github.com/rook/kubectl-rook-ceph/pkg/logging"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type fsStruct struct {
+	Name string
+	data []string
+}
+
+type subVolumeInfo struct {
+	svg string
+	fs  string
+}
+
+func List(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, includeStale bool) {
+
+	subvolumeNames := getK8sRefSubvolume(ctx, clientsets)
+	listCephFSSubvolumes(ctx, clientsets, operatorNamespace, clusterNamespace, includeStale, subvolumeNames)
+
+}
+
+// getk8sRefSubvolume returns the k8s ref for the subvolumes
+func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map[string]subVolumeInfo {
+	pvList, err := clientsets.Kube.CoreV1().PersistentVolumes().List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("Error fetching PVs: %v\n", err))
+	}
+	subvolumeNames := make(map[string]subVolumeInfo)
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI.VolumeAttributes["subvolumeName"] != "" {
+			subvolumeNames[pv.Spec.CSI.VolumeAttributes["subvolumeName"]] = subVolumeInfo{}
+		}
+	}
+	return subvolumeNames
+}
+
+// listCephFSSubvolumes list all the subvolumes
+func listCephFSSubvolumes(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, includeStale bool, subvolumeNames map[string]subVolumeInfo) map[string]subVolumeInfo {
+
+	// getFilesystem gets the filesystem
+	fsstruct := getFileSystem(ctx, clientsets, operatorNamespace, clusterNamespace)
+	staleSubvolumeNames := make(map[string]subVolumeInfo)
+	var svList string
+	if includeStale {
+		fmt.Println("Filesystem  Subvolume  SubvolumeGroup")
+	} else {
+		fmt.Println("Filesystem  Subvolume  SubvolumeGroup  State")
+	}
+	// this iterates over the filesystems and subvolumegroup to get the list of subvolumes that exist
+	for _, fs := range fsstruct {
+		// gets the subvolumegroup in the filesystem
+		subvolg := getSubvolumeGroup(ctx, clientsets, operatorNamespace, clusterNamespace, fs.Name)
+		for _, svg := range subvolg {
+			svList = exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "ls", fs.Name, svg.Name}, operatorNamespace, clusterNamespace, true, false)
+			subvol := unMarshaljson(svList)
+			if len(subvol) == 0 {
+				continue
+			}
+			// append the subvolume which doesn't have any snapshot attached to it.
+			for _, sv := range subvol {
```

Comment:
To simplify this `for` loop, here is my suggestion. I believe there was some misunderstanding of the `includeStale` flag and where to use `continue`, so hopefully this clarifies. I haven't tested this code, but hopefully it is clear.
```go
for _, sv := range subvol {
	// Assume the volume is stale unless proven otherwise
	stale := true
	// lookup for subvolume in list of the PV references
	_, ok := subvolumeNames[sv.Name]
	if ok || checkSnapshot(ctx, clientsets, operatorNamespace, clusterNamespace, fs.Name, sv.Name, svg.Name) {
		// The volume is not stale if a PV was found, or it has a snapshot
		stale = false
	} 
	status := "stale"
	if !stale {
		if includeStaleOnly {
			continue
		}
		status = "in-use"
	}
	subvolumeNames[sv.Name] = subVolumeInfo{fs.Name, svg.Name}
	fmt.Println(fs.Name, sv.Name, svg.Name, status)
}
```